substrate/primitives/trie/src/proof_size_extension.rs

```diff
@@ -37,3 +54,429 @@ impl ProofSizeExt {
 		self.0.estimate_encoded_size() as _
 	}
 }
+
+/// Proof size estimations as recorded by [`RecordingProofSizeProvider`].
+///
+/// Each item is the estimated proof size as observed when calling
+/// [`ProofSizeProvider::estimate_encoded_size`]. The items are ordered by t
+pub struct RecordedProofSizeEstimations(pub VecDeque<usize>);
+
+/// Inner structure of [`RecordingProofSizeProvider`].
+struct RecordingProofSizeProviderInner {
+	inner: Box<dyn ProofSizeProvider + Send + Sync>,
+	proof_size_estimations: Vec<Vec<usize>>,
```

Comment:
I think we can have a comment here explaining that the vector contains estimation results grouped by transaction.